### PR TITLE
=typ Examples how to stop actor with cleanup (#22776)

### PR DESCRIPTION
**Pull request by patriknw**
Continued from https://github.com/akka/akka/pull/23890

I have 
* rewritten tests to new style and testkits
* Changed to `Behaviors`
* Added documentation page

Please review @gosubpl 

Refs #22776

### CODE REVIEW

**Comment by gosubpl** [comment]
LGTM!
Thanks again @patriknw  for resurrecting this.

File: akka-actor-typed-tests/src/test/java/jdocs/akka/typed/GracefulStopDocTest.java [hunk 1]
```diff
@@ -0,0 +1,102 @@
...
+public class GracefulStopDocTest {
+
+    //#master-actor
+
+    public abstract static class JobControl {
+        // no instances of this class, it's only a name space for messages
+        // and static methods
+        private JobControl() {
+            // no instances of this class, it's only a name space for messages
+            // and static methods
```
Comment (johanandren) [inline akka-actor-typed-tests/src/test/java/jdocs/akka/typed/GracefulStopDocTest.java]: Good, but maybe one comment saying that is enough? ;)

File: akka-actor-typed-tests/src/test/java/jdocs/akka/typed/GracefulStopDocTest.java [hunk 2]
```diff
@@ -0,0 +1,102 @@
...
+                })
+                .onMessage(GracefulShutdown.class, (ctx, msg) -> {
+                    ctx.getSystem().log().info("Initiating graceful shutdown...");
+
+                    // perform graceful stop, executing cleanup before final system termination
+                    // behavior executing cleanup is passed as a parameter to Actor.stopped
+                    return Behaviors.stopped (Behaviors.onSignal((context, PostStop) -> {
+                            context.getSystem().log().info("Cleanup!");
+                            return Behaviors.same();
+                }));
```
Comment (johanandren) [inline akka-actor-typed-tests/src/test/java/jdocs/akka/typed/GracefulStopDocTest.java]: funky indentation

File: akka-actor-typed-tests/src/test/scala/akka/actor/typed/scaladsl/GracefulStopSpec.scala [hunk 3]
```diff
@@ -0,0 +1,72 @@
...
+                probe.ref ! Done
+                Behaviors.same
+            }
+          }
+        }
+
+      spawn(behavior)
+      probeCld1.expectMsg(Done)
+      probeCld2.expectMsg(Done)
+      probe.expectMsg(Done)
```
Comment (johanandren) [inline akka-actor-typed-tests/src/test/scala/akka/actor/typed/scaladsl/GracefulStopSpec.scala]: Are we verifying the children are stopped before the actor itself here? Isn't it better to use a single probe and distinct done-messages ("child", "child", "parent") then, to verify the children are stopped first?

File: akka-actor-typed-tests/src/test/scala/docs/akka/typed/GracefulStopDocSpec.scala [hunk 4]
```diff
@@ -0,0 +1,119 @@
...
+//#imports
+
+import akka.actor.typed.TypedAkkaSpecWithShutdown
+import akka.testkit.typed.TestKit
+
+object GracefulStopDocSpec {
+
+  //#master-actor
+
+  object MasterControlProgramActor {
```
Comment (johanandren) [inline akka-actor-typed-tests/src/test/scala/docs/akka/typed/GracefulStopDocSpec.scala]: Tron! :D

File: akka-actor-typed-tests/src/test/java/jdocs/akka/typed/GracefulStopDocTest.java (hunk 1 above)
Comment (gosubpl) [reply akka-actor-typed-tests/src/test/java/jdocs/akka/typed/GracefulStopDocTest.java]: https://github.com/akka/akka/commit/f3d11cbc4b91eb0dd3697f88f58c62de83113808 shall I drop the lower one?

File: akka-actor-typed-tests/src/test/java/jdocs/akka/typed/GracefulStopDocTest.java (hunk 2 above)
Comment (gosubpl) [reply akka-actor-typed-tests/src/test/java/jdocs/akka/typed/GracefulStopDocTest.java]: isn't this auto formatted?

File: akka-actor-typed-tests/src/test/scala/docs/akka/typed/GracefulStopDocSpec.scala (hunk 4 above)
Comment (gosubpl) [reply akka-actor-typed-tests/src/test/scala/docs/akka/typed/GracefulStopDocSpec.scala]: or Burroughs Mainframe reference :)

File: akka-actor-typed-tests/src/test/scala/akka/actor/typed/scaladsl/GracefulStopSpec.scala (hunk 3 above)
Comment (gosubpl) [reply akka-actor-typed-tests/src/test/scala/akka/actor/typed/scaladsl/GracefulStopSpec.scala]: yes

File: akka-actor-typed-tests/src/test/java/jdocs/akka/typed/GracefulStopDocTest.java (hunk 2 above)
Comment (gosubpl) [reply akka-actor-typed-tests/src/test/java/jdocs/akka/typed/GracefulStopDocTest.java]: ok, can be fixed

**Comment by gosubpl** [comment]
Since I cannot push to `akka/akka` :wink: @johanandren comments reflected here: https://github.com/gosubpl/akka/commit/0c86051b679c52db2147bc83a27cd1dc0a604c03

File: akka-docs/src/main/paradox/index-typed.md [hunk 5]
```diff
@@ -6,11 +6,12 @@
 
 * [actors](actors-typed.md)
 * [coexisting](coexisting.md)
+* [actor-discovery](actor-lifecycle-typed.md)
```
Comment (chbatey) [inline akka-docs/src/main/paradox/index-typed.md]: name should be actor-lifecycle

File: akka-actor-typed-tests/src/test/scala/docs/akka/typed/GracefulStopDocSpec.scala [hunk 6]
```diff
@@ -0,0 +1,119 @@
...
+    sealed trait JobControlLanguage
```
Comment (chbatey) [inline akka-actor-typed-tests/src/test/scala/docs/akka/typed/GracefulStopDocSpec.scala]: Someone been using a mainframe recently? :)

**Review by chbatey — APPROVED** [review]
LGTM with the one comment

**Review by gosubpl — APPROVED** [review]
LGTM

**Review by johanandren — APPROVED** [review]

**Comment by gosubpl** [comment]
Again, many thanks!

**Comment by patriknw** [comment]
You're welcome, but it's you that should get the thanks!

**Pull request merged**